File: 2022/haskell/day09.hs
-- Advent of Code 2022, day 9
--  Venanzio Capretta

module Main where

import System.Environment
import Data.List
import Data.Char

import Control.Applicative
import qualified Data.Map as M

import FunParser
import AoCTools

main :: IO ()
main = do
  args <- getArgs
  puzzle (head args)

puzzle :: String -> IO ()
puzzle fileName = do
  input <- readFile fileName
  let xs = parseAll pInput input
  putStrLn ("Part 1: " ++ show (part1 xs))
  putStrLn ("Part 2: " ++ show (part2 xs))

-- Parsing the input

data Direction = L | R | U | D
  deriving (Eq,Show)

pData :: Parser (Direction,Int)
pData = do dir <- (symbol "L" >> return L) <|>
                    (symbol "R" >> return R) <|>
                      (symbol "U" >> return U) <|>
                        (symbol "D" >> return D)
           steps <- natural
           return (dir,steps)
           
pInput :: Parser [(Direction,Int)]
pInput = many pData

-- Part 1

type Coords = (Int,Int)

move :: Coords -> Direction -> Coords
move (x,y) L = (x-1,y)
move (x,y) R = (x+1,y)
move (x,y) U = (x,y+1)
move (x,y) D = (x,y-1)

dist :: Coords -> Coords -> Int
dist (x1,y1) (x2,y2) = max (abs (x2-x1)) (abs (y2-y1))

follow :: Coords -> Coords -> Coords
follow c1@(x1,y1) c2@(x2,y2) =
  if dist c1 c2 > 1 then (x1 + signum (x2-x1), y1 + signum (y2-y1)) else c1

tailPos :: Coords -> Coords -> [(Direction,Int)] -> [Coords]
tailPos hd tl [] = []
tailPos hd tl ((_,0):steps) = tailPos hd tl steps
tailPos hd tl ((d,n):steps) =
  let hd' = move hd d
      tl' = follow tl hd'
      steps' = (d,n-1):steps
  in if tl' == tl then tailPos hd' tl' steps' else tl' : tailPos hd' tl' steps'

tailPositions :: Coords -> Coords -> [(Direction,Int)] -> [Coords]
tailPositions hd tl steps = tl : tailPos hd tl steps

part1 :: [(Direction,Int)] -> Int
part1 = length . nub . tailPositions (0,0) (0,0)

-- Part 2

ropeMove :: [Coords] -> Direction -> [Coords]
ropeMove (hd:knots) d = let hd' = move hd d in ropeFollow hd' knots

ropeFollow :: Coords -> [Coords] -> [Coords]
ropeFollow k0 [] = [k0]
ropeFollow k0 (k1:knots) = let k1' = follow k1 k0 in k0 : ropeFollow k1' knots

ropePos :: [Coords] -> [(Direction,Int)] -> [Coords]
ropePos rope [] = []
ropePos rope ((_,0):steps) = ropePos rope steps
ropePos rope ((d,n):steps) =
  let tl = last rope
      rope' = ropeMove rope d
      tl' = last rope'
      steps' = (d,n-1):steps
  in if tl' == tl then ropePos rope' steps' else tl' : ropePos rope' steps'

ropePositions :: [Coords] -> [(Direction,Int)] -> [Coords]
ropePositions rope steps = last rope : ropePos rope steps

part2 :: [(Direction,Int)] -> Int
part2 = length . nub . ropePositions (take 10 (repeat (0,0)))

File: 2022/python/day09.py
# Advent of Code 2022, day 9
#   Venanzio Capretta

print("Advent of Code 2022, Day 9")

f = open("../input09")
input = f.read().splitlines()
f.close()

# Parsing the input

def parse_move(s):
  (mv,steps) = s.split(' ',1)
  return (mv,int(steps))

moves = [parse_move(s) for s in input]

# Part 1

def distance(x1,y1,x2,y2):
  return max(abs(x1-x2),abs(y1-y2))

def sign(x):
  if x==0:
    return 0
  else:
    return x//abs(x)

def move(x,y,dir):
  if dir == 'U':
    return (x,y+1)
  elif dir == "D":
    return (x,y-1)
  elif dir == "R":
    return (x+1,y)
  elif dir == "L":
    return (x-1,y)
  else:
    return None

def follow(hx,hy,tx,ty):
  if distance(hx,hy,tx,ty) <= 1:
    return (tx,ty)
  else:
    return (tx+sign(hx-tx), ty+sign(hy-ty))

visited = {(0,0)}

(hx,hy) = (tx,ty) = (0,0)

for (dir,steps) in moves:
  for i in range(steps):
    (hx,hy) = move(hx,hy,dir)
    (tx,ty) = follow(hx,hy,tx,ty)
    visited.add((tx,ty))
  

print("Part 1: ")
print(len(visited))



# Part 2

print("Part 2: ")
  

